merge/merge.py

```diff
@@ -0,0 +1,83 @@
+from functools import partial, reduce
+from typing import Any, Callable, Optional, Sequence, Tuple
+
+
+def list_default(list1: list, list2: list) -> list:
+    # return reduce(lambda l1, l2: l1 + l2, [list1, list2])
+    return list1
+
+
+def dict_default(dict1: dict, dict2: dict, list_strategy: Callable) -> dict:
+    return merge(dict1, dict2, dict_strategy=dict_default, list_strategy=list_strategy)
```

Comment:
I don't understand this. So if they don't pass a dict_strategy to merge, then dict_strategy is supplied dict_default which is a function that references itself? I'll need an explanation on how this one works I'm not able to figure it out on my own. I can give more helpful feedback once I know how more of this works

Reply:
this was a late-night addition (proving that you should push commits after a certain time of the day...). I actually just removed it with the most recent commit, but it's doing the same thing (below). I wanted to parameterize the handling of dicts and lists because it allows someone to have more control over the specific behavior of a merge. for example, if you set `dict_strategy` to `lambda x, y: x`, the behavior of `merge` would be identical to the expression `{**dict1, **dict2}`, which is an idiomatic way of combining two dicts at the top level. Obviously having a strategy for handling primitive types would complete the control over merging, but I just haven't spent any time thinking of a non-intuitive way of dealing with that.

```python
def merge(
    dict1: dict, 
    dict2: dict, 
    dict_strategy: Callable = None,
    list_strategy: Callable = lambda x, y: x
    ) -> dict:
    """
    `merge` is a function to recursively, uhh, merge two dictionaries
    """
    if not dict_strategy:
        dict_strategy = partial(merge, list_strategy=list_strategy)

    keys = get_keys(dict1, dict2)

    return {
        key: merge_map(
            *get_values(dict1, dict2, key), 
            dict_strategy, 
            list_strategy)
        for key in keys
    }
```